File: 二分/33. 搜索旋转排序数组.py
# 假设按照升序排序的数组在预先未知的某个点上进行了旋转。
#
#  ( 例如，数组 [0,1,2,4,5,6,7] 可能变为 [4,5,6,7,0,1,2] )。
#
#  搜索一个给定的目标值，如果数组中存在这个目标值，则返回它的索引，否则返回 -1 。
#
#  你可以假设数组中不存在重复的元素。
#
#  你的算法时间复杂度必须是 O(log n) 级别。
#
#  示例 1:
#
#  输入: nums = [4,5,6,7,0,1,2], target = 0
# 输出: 4
#
#
#  示例 2:
#
#  输入: nums = [4,5,6,7,0,1,2], target = 3
# 输出: -1
#  Related Topics 数组 二分查找
#  👍 851 👎 0


# leetcode submit region begin(Prohibit modification and deletion)
class Solution(object):
    def search(self, nums, target):
        """
        :type nums: List[int]
        :type target: int
        :rtype: int
        """
        length = len(nums)
        l = 0
        r = length-1
        while l <= r:


            mid = (l +r)//2

            if nums[mid] == target:
                return mid
            if nums[0] <= nums[mid]: # 左边不包含旋转
                if nums[0] <= target <= nums[mid]:  # target在区间内
                    r = mid -1
                else:
                    l = mid +1
            else:                   #右边不包含旋转
                if nums[length-1] >= target >= nums[mid]:
                    l = mid +1
                else:
                    r = mid -1
        return  -1



ans  = Solution.search(None,[3,1], 1)
print(ans)

# leetcode submit region end(Prohibit modification and deletion)

File: 数组/33. 搜索旋转排序数组.java
package 数组;


// 实现了时间复杂度为O(logn)的算法，关键就是利用二分查找。
// 左右必有一边是绝对有序的。那么可以根据有序一边的极值来判断target在不在这一边，如果在，就继续在这一边找，如果不在，就在另一边找。
class Solution {
    public int search(int[] nums, int target) {
        int left = 0, right = nums.length - 1;
        return find(nums, target, left, right);

    }

    public int find(int[] nums, int target, int left, int right) {
        // 找不到
        if (left > right) {
            return -1;
        }
        int mid = (left + right) / 2;
        if (nums[mid] == target) {
            return mid;
        }
        // 如果左边有序
        if (nums[left] <= nums[mid]) {
            // 在范围内
            if (target >= nums[left] && target <= nums[mid]) {
                return find(nums, target, left, mid);
            } else {
                return find(nums, target, mid + 1, right);
                
            }
        }
        else { // 左边无序，右边有序
            if (target >= nums[mid] && target <= nums[right]) { // 看要不要找右边
                return find(nums, target, mid+1, right);
            } else {
                return find(nums, target, left, mid);
            }
        }
        
            
        

    }
}